rest-client-base/src/main/java/com/wultra/core/rest/client/base/DefaultRestClient.java

```diff
@@ -0,0 +1,700 @@
+/*
+ * Copyright 2020 Wultra s.r.o.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.wultra.core.rest.client.base;
+
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.fasterxml.jackson.databind.type.TypeFactory;
+import io.getlime.core.rest.model.base.request.ObjectRequest;
+import io.getlime.core.rest.model.base.response.ErrorResponse;
+import io.getlime.core.rest.model.base.response.ObjectResponse;
+import io.netty.channel.ChannelOption;
+import io.netty.handler.ssl.SslContext;
+import io.netty.handler.ssl.SslContextBuilder;
+import io.netty.handler.ssl.util.InsecureTrustManagerFactory;
+import org.springframework.core.ParameterizedTypeReference;
+import org.springframework.http.HttpHeaders;
+import org.springframework.http.MediaType;
+import org.springframework.http.ResponseEntity;
+import org.springframework.http.client.reactive.ReactorClientHttpConnector;
+import org.springframework.util.MultiValueMap;
+import org.springframework.web.reactive.function.BodyInserters;
+import org.springframework.web.reactive.function.client.*;
+import reactor.core.publisher.Mono;
+import reactor.netty.http.client.HttpClient;
+import reactor.netty.tcp.ProxyProvider;
+
+import javax.net.ssl.SSLException;
+import java.io.IOException;
+import java.lang.reflect.Type;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.util.function.Consumer;
+
+/**
+ * Default REST client implementation based on WebClient.
+ *
+ * @author Roman Strobl, roman.strobl@wultra.com
+ */
+public class DefaultRestClient implements RestClient {
+
+    private WebClient webClient;
+    private final RestClientConfiguration config;
+
+    /**
+     * Construct default REST client without any additional configuration.
+     * @param baseUrl Base URL.
+     * @throws RestClientException Thrown in case client initialization fails.
+     */
+    public DefaultRestClient(String baseUrl) throws RestClientException {
+        this.config = new RestClientConfiguration();
+        this.config.setBaseUrl(baseUrl);
+        // Use default WebClient settings
+        initializeWebClient();
+    }
+
+    /**
+     * Private constructor for builder.
+     * @param builder REST client builder.
+     * @throws RestClientException Thrown in case client initialization fails.
+     */
+    private DefaultRestClient(Builder builder) throws RestClientException {
```

Comment:
Do we want to introduce a specific constructor with the `RestClientConfiguration` parameter? The concern that I have is the following: We have a public `RestClientConfiguration` that you cannot use. We should either allow using it in the constructor or making it package-private.

Reply:
Good point. I think it is useful to add the constructor, so that e.g. JSON-serialized configuration can be easily configured.